### ISSUE TITLE
Theme Preview with no themes leads to a crash

### ISSUE THREAD

**Issue by TurdPooCharger** [astronautlevel2/Anemone3DS#168]
Hi again. 

If you have no themes in the **Themes** folder and press (Y) in  the theme selection menu, you crash.

[crash_dump_00000037.zip](https://github.com/astronautlevel2/Anemone3DS/files/2002084/crash_dump_00000037.zip)

Normally this doesn't happen because who uses Anemone without having any themes?

**Comment by astronautlevel2** [comment]
Do you just sit in your room and try to find bugs all day?

**Comment by TurdPooCharger** [comment]
It's raining where I'm from, and I can't do more important stuff at this moment, like yard work that my HoA has been giving my family and me a hard time about. I sure as heck ain't going out and risk catching the sniffles as much as I rather rack weeds and old grass for a fresh do over in the greenery department. But that's enough about me as this about Anemone we're focusing on. 

If you don't want any more bugs reported or would like them reported between timed and spaced successions for some room and fresh breather, I'll leave your GitHub be. I stated in the last submitted issue that I felt my reports were well past the point of welcoming; if I was in your shoes, I sure as hell would get sick of seeing me badgering you and your crew with seemingly non-ending streams of newly discovered bugs. That's stressful for a developer to bear; I rather not make anyone's day a bad one, especially a fellow programmer. 

As for finding bugs, some of these were discovered by accident due to previous tests. Frequently changing SD card build and Anemone setups produces a variety of conditions not normally seen in average use cases; it's not a surprise pitfalls not accounted for in Anemone's code base would show up for these weird scenarios.

I don't like writing these issues as I'm well aware how easy it is for human nature to form beef between people, but I reluctantly report anyway because I want Anemone to be the best 3ds theme manager that it can be. However, I also know online communication can came across wrong because written words don't carry context. If I've somehow offended you or someone in Anemone before, I apologies. I want to know what I did wrong so that I know how and what I can do better in messaging you guys.

Also, many 3DS users aren't concerned enough or are too scared sharing their bug discoveries because they don't want to put themselves in a position of getting ridiculed or be perceived as a noob. Some of them can't put into words explaining what the problems are they're seeing, either from a lack of fundamental understanding in 3DS homebrew or not having th vocabulary in describing the what, why, and how behind those problems. To them, you are a knowledgeable person of authority in 3DS homebrew(dom), and to come speak to you about something wrong with your app takes a lot guts on their part.

Normally, I would be working other 3DS projects like GM9 scripts but at this moment, I have coder's block. :frowning:

**Comment by astronautlevel2** [comment]
Most of the frustration comes from how long your posts typically are. A lot of your bug reports contain a huge amount of information which actually doesn't solve the problem, and just creates more noise for us when we're working on a solution. A perfect example of this was your report on the BGM loop bug, and even after I explained why your data were inaccurate, you continued on insisting that the `00000001` directory was the problem when really it was entirely unrelated. To a lesser extent, the logo bug exhibited the same problem.

A bit of it also comes from you going out of your way to find pretty much every single obscure way to break the program. Although you're definitely correct that these bugs should be solved, it becomes a bit overwhelming when all these obscure bugs that would never be encountered in regular use get in the way of us trying to solve either bigger problems or work on bigger features.

I'm not telling you to f off or anything, just maybe scale it back a bit and make your posts a bit more succinct.

**Comment by TurdPooCharger** [comment]
@astronautlevel2, the reason why my posts are long is due to making sure you guys get all the information in recreating my test setups as closely as possible, however miniscule or unimportant some of those details may seem. This is to ensure the bugs that were found have a better chance of being reproducible from your end so that you guys can fix and improve Anemone’s stability and operability. 

I hail from an engineering background where my fellow peers and those in related fields do not take lightly with poor explanations in problem solving, designing stuff, and describing processes. However, I admit the way I explain things was said to be more on the excessive side by my university friends and classmates, even by their standards… So I’m sorry for being very wordy. When you get hit with enough C’s in your classes because your technical papers don’t go into enough details, you soon end up learning how to write the way I do. 
***
***
I was going to let this one go after @LiquidFenrir managed to solve Anemone’s longstanding BGM loop bug with his clever 00 padding. Since you brought this up and you closed the discussion in its issue page, this is probably as good as any time in mentioning what I found about your theme manager.
***
- Anemone3DS handles theme shuffling differently than how Themely, CHMM2, and the stock 3DS theme manager does it.
***
Like I mentioned up front in that winding report I wrote, the files in the **000002cd** folder are always slightly changed when you delete it and let the 3DS rebuild it anew. This was the reason why I tried very hard in getting the crucial point across about using a seed folder to make the tests fair for all three managers. If that folder gets rebuilt but not replaced with the seed one, you won’t see the results I was getting because **00000001** changes by 2% between each rebuild and every other file undergoes similar changes. That seed folder was the reference to compare against the three managers’ outputs.

When I made that report, I had the shortcoming of not knowing what DIFF extraction and FBI’s Ext Save Data were until @CecilFF4 informed me in how to use them (thanks Cecil!). I didn’t understand at that time you were asking for the decrypted forms of those files, which was a big mistaken on my part. When I went back and did a hash check on the decrypted files from TEST No.3, this was what I found:
 
![decrypted comparison](https://user-images.githubusercontent.com/35666439/40095420-f64675c6-5899-11e8-9a15-8bd1b29f02b1.png)

Simply put, you were right that **00000001** wasn’t the one causing the BGM bug but rather a side effect of Anemone’s approach in producing those BgmCache files. It was bad interpretation on my part in believing that the first and most noticeably changed file was the root of it all.

I also did some comparison between the 3DS stock manager and CHMM2 with two free downloaded themes. CHMM had a theme dumper, making this comparison possible to do.

![3ds tm vs chmm2](https://user-images.githubusercontent.com/35666439/40095744-9a58b48e-589b-11e8-9182-f18d64d91b4c.png)

I don’t know how to read what was written in those BIN files, but I do know from observing at the macroscopic level, your theme manager’s approach to shuffling is unique. 

Even though the BGM loop bug has been fixed, if you or someone in your team is still interested in studying the data in those files from Test No.3, I’m still up for sharing what I’ve gathered.
